torch/csrc/jit/passes/onnx/fold_if_node.cpp

```diff
@@ -0,0 +1,249 @@
+  if (prev_node->kind() == onnx::If) {
+    int cond = 1 - (int)constantFoldedConditionValue(prev_node);
+    Block* block = prev_node->blocks()[cond];
+    // we are assuming that the number of block outputs is 1.
+    // TODO: make this more general by supporting the case
+    // of multiple outputs
+    prev_node = block->outputs()[0]->node();
+  }
+
+  if (prev_node->kind() == onnx::Constant) {
+    const at::Tensor& val = prev_node->t(attr::value);
+    return at::is_nonzero(val);
+  }
+
+  if (prev_node->kind() == onnx::Identity &&
+      prev_node->input()->node()->kind() == onnx::Constant) {
+    auto val = prev_node->input()->node()->t(attr::value);
+    return at::is_nonzero(val);
+  }
+
+  Node* compare_node = nullptr;
+  if (prev_node->kind() == onnx::Not) {
+    compare_node = prev_node->input()->node();
+  } else if (cast_node->inputs().size() > 0) {
+    compare_node = cast_node->input()->node();
+  }
+  TORCH_INTERNAL_ASSERT(compare_node != nullptr);
+  ScalarTypeAnalysisNodeForONNX(compare_node);
+  std::vector<at::Tensor> inputs;
+  for (size_t i = 0; i < compare_node->inputs().size(); i++) {
+    auto input_node = compare_node->inputs()[i]->node();
+    if (input_node->kind() == onnx::Constant) {
+      const at::Tensor& val = input_node->t(attr::value);
+      inputs.push_back(val);
+    } else { // input_node is either onnx::Size or onnx::ReduceProd
+      auto shape_node = input_node->input()->node();
+      auto shape =
+          shape_node->input()->type()->cast<TensorType>()->symbolic_sizes();
+
+      at::Tensor val;
+      if (input_node->kind() == onnx::Size) {
+        auto rank = shape.rank();
+        val = c10::scalar_to_tensor((int64_t)*rank);
+      } else if (input_node->kind() == onnx::ReduceProd) {
+        auto sizes = shape.sizes();
+        int64_t prod = 1;
+        for (int64_t i = 0; i < (int64_t)*shape.rank(); i++) {
+          auto dim = sizes.value()[i].static_size();
+          prod *= dim;
+        }
+        val = c10::scalar_to_tensor(prod);
+      }
+
+      inputs.push_back(val);
+    }
+  }
+
+  if (compare_node->kind() == onnx::Equal) {
+    if (prev_node->kind() == onnx::Not)
+      return !(at::equal(inputs[0], inputs[1]));
+    return at::equal(inputs[0], inputs[1]);
+  } else if (
+      (compare_node->kind() == onnx::Greater &&
+       prev_node->kind() != onnx::Not) ||
+      (prev_node->kind() == onnx::Not && compare_node->kind() == onnx::Less)) {
+    auto res = at::greater(inputs[0], inputs[1]);
+    if (prev_node->kind() == onnx::Not)
+      res = at::greater_equal(inputs[0], inputs[1]);
+    return at::is_nonzero(res);
+  } else if (
+      compare_node->kind() == onnx::Less ||
+      (prev_node->kind() == onnx::Not &&
+       compare_node->kind() == onnx::Greater)) {
+    auto res = at::less(inputs[0], inputs[1]);
+    if (prev_node->kind() == onnx::Not)
+      res = at::less_equal(inputs[0], inputs[1]);
+    return at::is_nonzero(res);
+  }
+  return false;
```

Comment:
If this condition cannot get true/false, we should go back and not doing const folding for if node.

Reply:
added assert